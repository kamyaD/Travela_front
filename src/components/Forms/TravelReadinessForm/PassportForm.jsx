import React, { PureComponent } from 'react';
import moment from 'moment';
import PropTypes from 'prop-types';
import axios from 'axios';
import toast from 'toastr';
import { isEmpty } from 'lodash';
import DocumentAPI from '../../../services/DocumentAPI';
import { FormContext, getDefaultBlanksValidatorFor } from '../FormsAPI';
import PassportDetails from'./FormFieldsets/passportDetails';
import './TravelReadiness.scss';
import SubmitArea from './FormFieldsets/SubmitArea';
import Preloader from '../../Preloader/Preloader';

class PassportForm extends PureComponent {
  constructor(props) {
    super(props);
    this.defaultState = {
      values: {
        name: '',
        passportNumber: '',
        nationality:'',
        dateOfBirth: '',
        dateOfIssue: '',
        placeOfIssue: '',
        expiryDate: ''
      },
      image: '',
      documentUploaded: false,
      uploadingDocument: false,
      errors: {},
      hasBlankFields: true,
    };
    this.state = {...this.defaultState};
  }


  componentWillReceiveProps(nextProps, nextContext) {
    const { errors, document } = nextProps;
    const { document: currentDocument } = this.props;
    if (isEmpty(currentDocument) && !isEmpty(document)) {
      const { passportNumber, name, nationality, dateOfBirth, dateOfIssue, placeOfIssue, expiryDate, cloudinaryUrl } = document.data;
      this.newState = {
        values: {
          name: name,
          passportNumber,
          nationality,
          dateOfBirth: moment(dateOfBirth),
          dateOfIssue: moment(dateOfIssue),
          placeOfIssue,
          expiryDate: moment(expiryDate)
        },
        id: document.id,
        image: cloudinaryUrl,
        imageChanged: false,
      };
      this.setState({errors, ...this.newState });
    }
    this.setState({ errors });
  }

  componentWillUnmount() {
    const { fetchUserData, user } = this.props;
    fetchUserData(user.currentUser.userId);
  }

  onCancel = (event) => {
    event.preventDefault();
    const { closeModal } = this.props;
    closeModal();
  };

  handleSubmit = async event => {
    event.preventDefault();
    const { values, image, errors, documentUploaded, id } = this.state;
    const { createTravelReadinessDocument, editTravelReadinessDocument, modalType } = this.props;
    const { dateOfBirth, dateOfIssue, expiryDate } = values;

    if( documentUploaded){

      const newValues = {
        ...values,
        dateOfBirth: moment(dateOfBirth).format('YYYY/MM/DD'),
        dateOfIssue:moment(dateOfIssue).format('YYYY/MM/DD'),
        expiryDate: moment(expiryDate).format('YYYY/MM/DD')
      };

      if (modalType === 'edit passport') {
        editTravelReadinessDocument('passport', newValues, id);
      } else {
        createTravelReadinessDocument('passport', newValues);
      }
    }else {
      if (image) {
        const fd = new FormData();
        fd.append('file', image);
        fd.append('upload_preset', process.env.REACT_APP_PRESET_NAME);

        try {
          delete axios.defaults.headers.common['Authorization'];
          this.setState({uploadingDocument: true});
          const imageData = await axios.post(process.env.REACT_APP_CLOUNDINARY_API, fd);
          const {data: {url}} = imageData;
          this.setState({
            documentUploaded: true,
            uploadingDocument: false,
            values: {...values, cloudinaryUrl: url}
          });
          DocumentAPI.setToken();
          const passportValues = {
            ...values, 
            cloudinaryUrl: url,
            dateOfBirth: moment(dateOfBirth).format('YYYY/MM/DD'),
            dateOfIssue:moment(dateOfIssue).format('YYYY/MM/DD'),
            expiryDate: moment(expiryDate).format('YYYY/MM/DD')
          };

          if (modalType === 'edit passport') {
            editTravelReadinessDocument('passport',{ ...passportValues }, id);
          } else {
            createTravelReadinessDocument('passport',{ ...passportValues });
          }
        } catch (e) {
          toast.error('Error uploading passport. Please try again!');
        }
      } else {
        this.setState({errors: {...errors, cloudinaryUrl: 'Please upload a passport'}});
      }
    }
  };

  handleUpload=(e) => {
    const { errors } = this.state;
    const file = e.target.files[0];
    const tenMBS = Math.pow(10,7);
    if (file.size > tenMBS){
      return toast.error('Incorrect file size uploaded');
    }
    this.setState({
      image: e.target.files[0],
      imageChanged: true,
      documentUploaded: false,
      errors: { ...errors, cloudinaryUrl: ''}});
  };

  render() {
    const {errors, values, hasBlankFields,uploadingDocument, imageChanged } = this.state;
    const { document, modalType, fetchingDocument } = this.props;
    return (
      <div>
        {fetchingDocument ? <Preloader /> : (
          <FormContext
            targetForm={this}
            errors={errors}
            validatorName="myCustomValidator" // uses FormAPI's default validator if this is undefined
            values={values}>
            <form className="passport-form" onSubmit={this.handleSubmit}>
              <PassportDetails 
                handleUpload={this.handleUpload} 
                errors={errors} 
                document={document}
                modalType={modalType}
              />
              <SubmitArea
                onCancel={this.onCancel} hasBlankFields={hasBlankFields && !imageChanged}
                send={modalType === 'edit passport' ? 'Save Changes' : 'Add Passport'}
                loading={uploadingDocument} />
            </form>
          </FormContext>)}
      </div>
    );
  }

}

PassportForm.propTypes = {
  createTravelReadinessDocument: PropTypes.func.isRequired,
  errors: PropTypes.shape({}).isRequired,
  fetchUserData: PropTypes.func.isRequired,
  user: PropTypes.object.isRequired,
  closeModal: PropTypes.func.isRequired,
  editTravelReadinessDocument: PropTypes.func,
  document: PropTypes.object.isRequired,
  modalType: PropTypes.string,
  fetchingDocument: PropTypes.bool
};

PassportForm.defaultProps = {
  editTravelReadinessDocument: () => {},
  modalType: '',
  fetchingDocument: false
};

export default PassportForm;
