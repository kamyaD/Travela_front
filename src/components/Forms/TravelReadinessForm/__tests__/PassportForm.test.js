import toast from 'toastr';
import moxios from 'moxios';
import React from 'react';
import PassportForm from '../PassportForm';
import PassportDetailsFieldSet from '../FormFieldsets/passportDetails';

toast.success = jest.fn();
toast.error = jest.fn();

describe('<PassportForm/>',  () => {

  let wrapper, createTravelReadinessDocument, editTravelReadinessDocument;
  createTravelReadinessDocument = jest.fn();
  editTravelReadinessDocument = jest.fn();

  const props = {
    createTravelReadinessDocument: jest.fn(),
    editTravelReadinessDocument: jest.fn(),
    fetchingDocument: false,
    isLoading: false,
    document: {},
    errors: {},
    fetchUserData: jest.fn(),
    users: [],
    user: {},
    closeModal: jest.fn(),
    onCancel: jest.fn()
  };

  const document = {
    name: 'Mike',
    passportNumber : 'ABC123aemlbczy85',
    nationality : 'Kenyan ',
    placeOfIssue : 'Kenya',
    cloudinaryUrl : 'https://farm4.staticflickr.com/3894/15008518202_c265dfa55f_h.jpeg'
  };


  const passportFile = new Blob(['Passport file'], {type: 'application/pdf', name: 'Filename.pdf'});

  const fileSize = passportFile.size * 10;

  const event = {
    target: {
      files: [
        passportFile
      ]
    }
  };

  beforeEach(() => {
    wrapper = mount(<PassportForm {...props} />);
    moxios.install();
  });

  afterEach(() => {
    moxios.uninstall();
  });

  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('renders the edit passport document modal', () => {
    const wrapperWithPassportField = mount(
      <PassportForm 
        {...props} 
        documentType="passport"
        modalType="edit passport"
      />
    );
    expect(
      wrapperWithPassportField.find(PassportDetailsFieldSet))
      .toHaveLength(1);
  });

  it('renders passport  modal with button text \'Add Passport\'', () => {
    const wrapperWithOtherDocumentField = mount(
      <PassportForm {...props} documentType="passport" />
    );
    expect(
      wrapperWithOtherDocumentField.find('button#submit').text()).toEqual('Add Passport');
  });

  it('renders edit passport  modal with button text \'Save Changes\'', () => {
    const wrapperWithOtherDocumentField = mount(
      <PassportForm {...props} documentType="passport" modalType="edit passport" />
    );
    expect(
      wrapperWithOtherDocumentField.find('button#submit').text()).toEqual('Save Changes');
  });

  it('should upload a file', () => {
    wrapper.find('#select-file').simulate('change', event);
    expect(wrapper.state().image).toEqual(passportFile);
  });

  it('should upload a file for edit', () => {
    const newProps = {
      ...props,
      modalType: 'edit passport'
    };
    const wrapper = mount(<PassportForm {...newProps} />);
    wrapper.find('#select-file').simulate('change', event);
    expect(wrapper.state().image).toEqual(passportFile);
  });

  it('should not submit without document', () => {
    wrapper.setState({values: {...document}});

    process.env.REACT_APP_CLOUNDINARY_API = 'https://mock-passport-cloudinary-api-succeed';

    const cloudinaryResponse = {
      url: 'secure url'
    };

    moxios.stubRequest(process.env.REACT_APP_CLOUNDINARY_API, {
      status: 200,
      response: cloudinaryResponse
    });

    wrapper.find('.passport-form').simulate('submit');

    moxios.wait(() => {
      expect(wrapper.state().documentUploaded).toBeFalsy();
      expect(toast.error).toHaveBeenCalledWith('Please upload a passport.');
    });
  });

  it('should call the handleSubmit method for passport edits', () => {
    const newProps = {
      ...props,
      modalType: 'edit passport',
      documentType: 'passport'
    };
    const wrapperWithPassportField = mount(
      <PassportForm {...newProps} />
    );

    const instance = wrapperWithPassportField.instance();
    jest.spyOn(instance, 'handleSubmit');
    let documentUploaded = false;
    wrapperWithPassportField.find('form').simulate('submit', {
      preventDefault: () => {
        documentUploaded = true;
      },
    });
    expect(documentUploaded).toBe(true);
  });

  it('should submit if all the data is valid',  () => {
    wrapper.setState({values: {...document}});
    wrapper.find('#select-file').simulate('change', event);


    process.env.REACT_APP_CLOUNDINARY_API = 'https://mock-passport-cloudinary-api-succeed';

    const cloudinaryResponse = {
      url: 'secure url'
    };

    moxios.stubRequest(process.env.REACT_APP_CLOUNDINARY_API, {
      status: 200,
      response: cloudinaryResponse
    });

    wrapper.find('.passport-form').simulate('submit');

    moxios.wait(() => {
      expect(wrapper.state().documentUploaded).toBeTruthy();
      expect(createTravelReadinessDocument).toHaveBeenCalledWith('passport', document);
    });
  });

  it('should not upload an image if it is already uploaded', () => {
    wrapper.setState({values: {...document}});
    wrapper.find('#select-file').simulate('change', event);


    process.env.REACT_APP_CLOUNDINARY_API = 'https://mock-passport-cloudinary-api-succeed';

    const cloudinaryResponse = {
      url: 'secure url'
    };

    moxios.stubRequest(process.env.REACT_APP_CLOUNDINARY_API, {
      status: 200,
      response: cloudinaryResponse
    });

    wrapper.setState({documentUploaded: true});
    wrapper.find('.passport-form').simulate('submit');
    expect(props.createTravelReadinessDocument).toHaveBeenCalled();
  });

  it('should display errors when received from the api', () => {
    wrapper.setProps({ errors: { cloudinaryUrl: 'Please provide a passport'}});
    expect(wrapper.state().errors).toEqual({cloudinaryUrl: 'Please provide a passport'});
  });

  it('should toast an error if the passport file does not upload', () => {
    wrapper.setState({values: {...document}});
    wrapper.find('#select-file').simulate('change', event);

    process.env.REACT_APP_CLOUNDINARY_API = 'https://mock-passport-cloudinary-api-succeed';

    const cloudinaryResponse = {
      url: 'secure url'
    };

    moxios.stubRequest(process.env.REACT_APP_CLOUNDINARY_API, {
      status: 404,
      response: cloudinaryResponse
    });

    wrapper.setState({documentUploaded: false});
    wrapper.find('.passport-form').simulate('submit');
    expect(toast.error).toHaveBeenCalledTimes(0);
  });

  it('toasts an error if cloudinary returns an error', () => {
    moxios.stubRequest(process.env.REACT_APP_CLOUNDINARY_API, { status: 500});
    wrapper.find('#select-file').simulate('change', event);
  });

  it('should edit if all the data is valid',  () => {
    const newProps = {
      ...props,
      modalType: 'edit passport'
    };

    wrapper = mount(<PassportForm {...newProps} />);
    wrapper.setState({values: {...document}});
    wrapper.find('#select-file').simulate('change', event);


    process.env.REACT_APP_CLOUNDINARY_API = 'https://mock-passport-cloudinary-api-succeed';

    const cloudinaryResponse = {
      url: 'secure url'
    };

    moxios.stubRequest(process.env.REACT_APP_CLOUNDINARY_API, {
      status: 200,
      response: cloudinaryResponse
    });

    wrapper.find('.passport-form').simulate('submit');

    moxios.wait(() => {
      expect(wrapper.state().documentUploaded).toBeTruthy();
      expect(createTravelReadinessDocument).toHaveBeenCalledWith('passport', document);
    });
  });

  it('should not upload an image for edit if it is already uploaded', () => {
    const newProps = {
      ...props,
      modalType: 'edit passport'
    };
    wrapper = mount(<PassportForm {...newProps} />);
    wrapper.setState({values: {...document}});
    wrapper.find('#select-file').simulate('change', event);


    process.env.REACT_APP_CLOUNDINARY_API = 'https://mock-passport-cloudinary-api-succeed';

    const cloudinaryResponse = {
      url: 'secure url'
    };

    moxios.stubRequest(process.env.REACT_APP_CLOUNDINARY_API, {
      status: 200,
      response: cloudinaryResponse
    });

    wrapper.setState({documentUploaded: true});
    wrapper.find('.passport-form').simulate('submit');
    expect(props.editTravelReadinessDocument).toHaveBeenCalled();
  });

  it('should display an error when there is no value in input', () => {
    const formWrapper = mount(
      <PassportForm {...props} documentType="passport" />
    );
    let documentUploaded = false;
    formWrapper.find('form').simulate('submit', {
      preventDefault: () => {
        documentUploaded = true;
      },
    });
    expect(documentUploaded).toBe(true);
  });

  it('should receive new props for the input fields if user wants to edit a document', () => {
    const newProps = {
      ...props,
      editTravelReadinessDocument: jest.fn(),
      document: {
        data: {
          passportNumber: '6yy2',
          name: 'emeka',
          nationality: 'Nigerian',
          placeOfIssue: 'Naija', 
          cloudinaryUrl: 'image.com'
        }
      },
      currentDocument: {},
    };
    wrapper = mount(<PassportForm {...newProps} />);
    wrapper.setState({values: {...document}});
  });

  it('should cancel if user no longer wants to edit a document', () => {
    const newProps = {
      ...props,
      editTravelReadinessDocument: jest.fn(),
      document: {
        data: {
          passportNumber: '6yy2',
          name: 'emeka',
          nationality: 'Nigerian',
          placeOfIssue: 'Naija', 
          cloudinaryUrl: 'image.com'
        }
      },
      currentDocument: {},
    };
    wrapper = mount(<PassportForm {...newProps} />);
    wrapper.setState({values: {...document}});
  });

  it('should display errors when received from the api', () => {
    wrapper.setProps({ errors: { cloudinaryUrl: 'Please provide a document'}});
    expect(wrapper.state().errors).toEqual({cloudinaryUrl: 'Please provide a document'});
  });
});
