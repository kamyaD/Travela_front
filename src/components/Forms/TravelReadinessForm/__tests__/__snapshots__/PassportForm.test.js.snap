// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PassportForm/> renders correctly 1`] = `
<div>
  <form
    className="travel-document-form"
    onSubmit={[Function]}
  >
    <div>
      <fieldset>
        <div
          className="input-group"
        >
          <div
            className="form-input fluid "
          >
            <label
              htmlFor="name"
            >
              Name as Seen On Passport
              <span
                style={
                  Object {
                    "color": "red",
                  }
                }
              >
                 * 
              </span>
            </label>
            <input
              autoComplete="off"
              className=" fluid"
              label="Name as Seen On Passport"
              name="name"
              onBlur={[Function]}
              onChange={[Function]}
              type="text"
              value=""
            />
            <span
              className="error"
            />
          </div>
          <div
            className="form-input  "
          >
            <label
              htmlFor="passportNumber"
            >
              Passport Number
              <span
                style={
                  Object {
                    "color": "red",
                  }
                }
              >
                 * 
              </span>
            </label>
            <input
              autoComplete="off"
              className=" "
              label="Passport Number"
              name="passportNumber"
              onBlur={[Function]}
              onChange={[Function]}
              type="text"
              value=""
            />
            <span
              className="error"
            />
          </div>
          <div
            className="form-input  "
          >
            <label
              htmlFor="nationality"
            >
              Nationality
              <span
                style={
                  Object {
                    "color": "red",
                  }
                }
              >
                 * 
              </span>
            </label>
            <input
              autoComplete="off"
              className=" "
              label="Nationality"
              name="nationality"
              onBlur={[Function]}
              onChange={[Function]}
              type="text"
              value=""
            />
            <span
              className="error"
            />
          </div>
          <div
            className="form-input  "
          >
            <label
              htmlFor="dateOfBirth"
            >
              Date of Birth
              <span
                style={
                  Object {
                    "color": "red",
                  }
                }
              >
                 * 
              </span>
            </label>
            <div
              className="date-wrapper "
              id="dateOfBirth_date"
            >
              <div>
                <div
                  className="react-datepicker-wrapper"
                >
                  <div
                    className="react-datepicker__input-container"
                  >
                    <input
                      autoComplete="off"
                      className=""
                      disabled={false}
                      name="dateOfBirth"
                      onBlur={[Function]}
                      onChange={[Function]}
                      onClick={[Function]}
                      onFocus={[Function]}
                      onKeyDown={[Function]}
                      placeholder="MM/DD/YYYY"
                      readOnly={false}
                      type="text"
                      value=""
                    />
                  </div>
                </div>
              </div>
              <img
                alt="cal"
                className="calendar-icon"
                src="calendar_icon.svg"
              />
            </div>
            <span
              className="error"
            />
          </div>
          <div
            className="form-input  "
          >
            <label
              htmlFor="dateOfIssue"
            >
              Date of Issue
              <span
                style={
                  Object {
                    "color": "red",
                  }
                }
              >
                 * 
              </span>
            </label>
            <div
              className="date-wrapper "
              id="dateOfIssue_date"
            >
              <div>
                <div
                  className="react-datepicker-wrapper"
                >
                  <div
                    className="react-datepicker__input-container"
                  >
                    <input
                      autoComplete="off"
                      className=""
                      disabled={false}
                      name="dateOfIssue"
                      onBlur={[Function]}
                      onChange={[Function]}
                      onClick={[Function]}
                      onFocus={[Function]}
                      onKeyDown={[Function]}
                      placeholder="MM/DD/YYYY"
                      readOnly={false}
                      type="text"
                      value=""
                    />
                  </div>
                </div>
              </div>
              <img
                alt="cal"
                className="calendar-icon"
                src="calendar_icon.svg"
              />
            </div>
            <span
              className="error"
            />
          </div>
          <div
            className="form-input  "
          >
            <label
              htmlFor="placeOfIssue"
            >
              Place of Issue
              <span
                style={
                  Object {
                    "color": "red",
                  }
                }
              >
                 * 
              </span>
            </label>
            <input
              autoComplete="off"
              className=" "
              label="Place of Issue"
              name="placeOfIssue"
              onBlur={[Function]}
              onChange={[Function]}
              type="text"
              value=""
            />
            <span
              className="error"
            />
          </div>
          <div
            className="form-input  "
          >
            <label
              htmlFor="expiryDate"
            >
              Expiry Date
              <span
                style={
                  Object {
                    "color": "red",
                  }
                }
              >
                 * 
              </span>
            </label>
            <div
              className="date-wrapper "
              id="expiryDate_date"
            >
              <div>
                <div
                  className="react-datepicker-wrapper"
                >
                  <div
                    className="react-datepicker__input-container"
                  >
                    <input
                      autoComplete="off"
                      className=""
                      disabled={false}
                      name="expiryDate"
                      onBlur={[Function]}
                      onChange={[Function]}
                      onClick={[Function]}
                      onFocus={[Function]}
                      onKeyDown={[Function]}
                      placeholder="MM/DD/YYYY"
                      readOnly={false}
                      type="text"
                      value=""
                    />
                  </div>
                </div>
              </div>
              <img
                alt="cal"
                className="calendar-icon"
                src="calendar_icon.svg"
              />
            </div>
            <span
              className="error"
            />
          </div>
        </div>
      </fieldset>
    </div>
    <div
      className="travel-document-select-file"
    >
      <p>
        Attach File
      </p>
      <div
        className="document-input__input-container"
      >
        <label
          className="document-input__input-container__prompts"
          htmlFor="select-file"
        >
          <img
            alt="Document Upload"
            className="document-input__input-container__prompts__img"
            src="document-upload-blue.svg"
          />
          <div
            className="document-input__input-container__prompts__text"
          >
            <p>
              Drag file here or choose from computer
            </p>
          </div>
          <input
            id="select-file"
            onChange={[Function]}
            type="file"
          />
        </label>
      </div>
    </div>
    <hr />
    <div
      className="travel-document-submit-area"
    >
      <fieldset>
        <div
          className="submit-area"
        >
          <button
            className="bg-btn bg-btn--inactive"
            id="cancel"
            onClick={[Function]}
            type="button"
          >
            Cancel
          </button>
          <button
            className="bg-btn bg-btn--active"
            disabled={true}
            id="submit"
            type="submit"
          >
            <undefined />
          </button>
        </div>
      </fieldset>
    </div>
  </form>
</div>
`;
